engine/runtime/src/test/java/org/enso/interpreter/test/BinaryDispatchTest.java

```diff
@@ -0,0 +1,181 @@
+                    Fraction r q
+
+                    --- self that:R =
+                        s = self.r/self.q
+                        t = that.r/that.q
+                        (s.abs - t.abs).abs
+
+                R.from (that:Z) = R.Fraction that.n 1
+
+                """,
+                "prelude.enso")
+            .build();
+    module = ctx.eval(prelude);
+  }
+
+  @AfterClass
+  public static void closeCtx() {
+    ctx.close();
+  }
+
+  @Test
+  public void verifyZ() {
+    var six = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 6");
+    var ten = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 10");
+
+    var diff1 = six.invokeMember("---", ten);
+    assertEquals(diff1.asInt(), 4);
+  }
+
+  @Test
+  public void verifyR() {
+    var half = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 1 2");
+    var oneEight = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 1 8");
+
+    var diff1 = half.invokeMember("---", oneEight);
+    assertEquals(diff1.asDouble(), 3.0 / 8, 0.01);
+  }
+
+  @Test
+  public void thatArgumentIsConverted() {
+    var half = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 1 2");
+    var two = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 2");
+
+    var diff1 = half.invokeMember("---", two);
+    assertEquals(diff1.asDouble(), 3.0 / 2, 0.01);
+  }
+
+  @Test
+  public void selfArgumentIsConverted() {
+    var one = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 1");
+    var threeHalfs = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 3 2");
+
+    var diff1 = one.invokeMember("---", threeHalfs);
+    assertEquals(diff1.asDouble(), 1.0 / 2, 0.01);
+  }
+
+  @Test
+  public void staticVerifyZ() {
+    var zOperator = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.---");
+    assertTrue("It's executable", zOperator.canExecute());
+
+    var six = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 6");
+    var ten = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 10");
+
+    var diff1 = zOperator.execute(six, ten);
+    assertEquals(diff1.asInt(), 4);
+  }
+
+  @Test
+  public void staticVerifyR() {
+    var rOperator = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.---");
+    assertTrue("It's executable", rOperator.canExecute());
+
+    var half = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 1 2");
+    var oneEight = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.Fraction 1 8");
+
+    var diff1 = rOperator.execute(half, oneEight);
+    assertEquals(diff1.asDouble(), 3.0 / 8, 0.01);
+  }
+
+  @Test
+  @Ignore // PENDING: #8805
+  public void staticWithRFirstArgumentIsConverted() {
+    var rOperator = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "R.---");
+
+    var two = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "Z.Number 2");
```

Comment:
Once #8805 is implemented, the `execute(two, half)` shall _convert Z_ into `R` first and use the converted value as `self` argument. Right now no conversion happens and `Z` is used as `self` which crashes the execution.